api/v1/api_test.go

```diff
@@ -0,0 +1,196 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package v1
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+
+	dto "github.com/prometheus/client_model/go"
+
+	"github.com/go-kit/kit/log"
+	"github.com/golang/protobuf/proto"
+	"github.com/prometheus/pushgateway/storage"
+)
+
+var (
+	logger    = log.NewNopLogger()
+	testFlags = map[string]string{
+		"flag1": "value1",
+		"flag2": "value2",
+		"flag3": "value3",
+	}
+	testBuildInfo = map[string]string{
+		"build1": "value1",
+		"build2": "value2",
+		"build3": "value3",
+	}
+
+	mf1 = &dto.MetricFamily{
+		Name: proto.String("mf1"),
+		Type: dto.MetricType_UNTYPED.Enum(),
+		Metric: []*dto.Metric{
+			{
+				Label: []*dto.LabelPair{
+					{
+						Name:  proto.String("instance"),
+						Value: proto.String("instance2"),
+					},
+					{
+						Name:  proto.String("job"),
+						Value: proto.String("job1"),
+					},
+				},
+				Untyped: &dto.Untyped{
+					Value: proto.Float64(-3e3),
+				},
+			},
+		},
+	}
+
+	grouping1 = map[string]string{
+		"job":      "job1",
+		"instance": "instance1",
+	}
+)
+
+func compareMaps(mainMap map[string]interface{}, testMap map[string]string) bool {
+	if len(mainMap) != len(testMap) {
+		fmt.Print(len(mainMap), len(testMap))
+		return false
+	}
+
+	for key, _ := range mainMap {
+		if mainMap[key].(string) != testMap[key] {
+			fmt.Print("Dagabaazi")
+			return false
+		}
+	}
+	return true
+}
+
+func TestStatusAPI(t *testing.T) {
+	dms := storage.NewDiskMetricStore("", 100*time.Millisecond, nil, logger)
+	TestApi := New(logger, dms, testFlags, testBuildInfo)
+
+	req, err := http.NewRequest("GET", "http://example.org/", &bytes.Buffer{})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	w := httptest.NewRecorder()
+
+	testResponse := response{}
+	TestApi.status(w, req)
+	json.Unmarshal(w.Body.Bytes(), &testResponse)
+	jsonData := testResponse.Data.(map[string]interface{})
+	responseFlagData := jsonData["flags"].(map[string]interface{})
+	responseBuildInfo := jsonData["build_information"].(map[string]interface{})
+
+	if expected, got := http.StatusOK, w.Code; expected != got {
+		t.Errorf("Wanted status code %v, got %v.", expected, got)
+	}
+
+	if !compareMaps(responseFlagData, testFlags) {
+		t.Errorf("Wanted following flags %v, got %v.", testFlags, responseFlagData)
+	}
+
+	if !compareMaps(responseBuildInfo, testBuildInfo) {
+		t.Errorf("Wanted following build info %v, got %v.", testBuildInfo, responseBuildInfo)
+	}
+}
+
+func TestMetricsAPI(t *testing.T) {
```

Comment:
@beorn7 I implemented TestMetricsApi but don't know how vigorous testing should be done
Can you help me with the same

Reply:
As commented separately, having a Histogram and or Summary in it, would be good. Since it is reflected in other comments, I'll resolve this one.